File: 54_spiral_matrix/src/Main.java
import java.util.ArrayList;
import java.util.List;

public class Main {
  class Solution {
    public static List<Integer> spiralOrder(int[][] matrix) {
      List<Integer> result = new ArrayList<>();
      if (matrix.length == 0) {
        return result;
      }

      int m = matrix[0].length - 1;
      int n = matrix.length - 1;
      int top = 0;
      int left = 0;
      int right = m;
      int bottom = n;

      while(top <= bottom && left <= right) {
        // go from left to right
        for (int i = left; i <= right; i++) {
          result.add(matrix[top][i]);
        }
        top++;

        // go from top right to bottom
        for (int i = top; i <= bottom; i++) {
          result.add(matrix[i][right]);
        }
        right--;

        // go from bottom right to left
        for (int i = right; i >= left; i--) {
          result.add(matrix[bottom][i]);
        }
        bottom--;

        // go from bottom left to top
        for (int i = bottom; i >= top; i--) {
          result.add(matrix[i][left]);
        }
        left++;
      }

      return result;
    }
  }

  public static void main(String[] args) {
    int[][] matrix = {
            {1,2,3},
            {4,5,6},
            {7,8,9}
    };
    System.out.println(Solution.spiralOrder(matrix));
  }
}

File: 54_spiral_matrix/main.go
package main

import (
	"fmt"
)

/**
iteration 1:
k = 0, i= 0, j =0
i = 0, j = 0 -> 3, k = 0
i = 0 -> 3, j = 3, k= 0
i = 3, j = 2 -> 0, k = 0
i = 2 -> 0, j = 0, k = 0

iteration 2:
k = 1
i =
**/

func spiralOrder(matrix [][]int) []int {
	var result []int
	if len(matrix) == 0 {
		return result
	}

	var i, j int
	N := len(matrix[0]) * len(matrix)
	k := 0

	for len(result) < N && i < len(matrix) && j < len(matrix[0]) {
		for ; len(result) < N && j < len(matrix[0])-k; j++ {
			result = append(result, matrix[i][j])
		}
		j--
		i++

		for ; len(result) < N && i < len(matrix)-k; i++ {
			result = append(result, matrix[i][j])
		}
		i--
		j--

		for ; len(result) < N && j >= k; j-- {
			result = append(result, matrix[i][j])
		}
		j++
		i--

		for ; len(result) < N && i > k; i-- {
			result = append(result, matrix[i][j])
		}

		i++
		j++
		k++
	}

	return result
}

func main() {
	matrix1 := [][]int{
		{1, 2, 3, 4},
		{5, 6, 7, 8},
		{9, 10, 11, 12},
	}

	fmt.Printf("%v", spiralOrder(matrix1))
	// expect:
	// [1, 2, 3, 6, 9, 8, 7, 4, 5]
}
